crypto/o_time.c

```diff
@@ -358,3 +358,19 @@ static void julian_to_date(long jd, int *y, int *m, int *d)
     *m = j + 2 - (12 * L);
     *y = 100 * (n - 49) + i + L;
 }
+
+
+int OPENSSL_timegm(struct tm *tm, time_t *t)
+{
+    /* Figures it out without mucking the environment */
+    long basetime, currtime;
+
+    basetime = date_to_julian(1970, 1, 1);
+    currtime = date_to_julian(tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday);
+
+    if (t != NULL)
+        *t = (((currtime - basetime) * 24 + tm->tm_hour) * 60 + tm->tm_min) * 60 + tm->tm_sec;
```

Comment:
@bernd-edlinger, this function is only necessary for those who need/want to get a time_t from an ASN1_TIME; as you correctly point out, certificates can be created beyond the time_t's 2038 limitation, but not by using time_t alone, since that's a 32-bit time_t limitation. You have to use something other than ASN1_TIME_set() as you point out. However, if you are on a 64-bit platform, then you can easily use ASN1_TIME_set() and ASN1_TIME_get() without any worry.

Also note that ASN1_TIME_get() can return a **struct tm** if you don't want to use time_t. (Yes, this blows away my orthogonality argument - but there's enough infrastructure to make it trivial to implement setting an ASN1_TIME via **struct tm**.)

That being said, this could be forced to be a 64-bit calculation, but that doesn't necessarily help 32-bit systems using time_t in the end.

@levitte the effort has been spent; it's actually quite simple with the julian functions.

I also consider these time wrapper functions to be similar to the OPENSSL_malloc()/OPENSSL_free() wrappers or the ASYNC wrappers around pthreads and the Windows APIs: an OS abstraction layer.